ci.nix

```diff
@@ -33,54 +39,56 @@ dimension "Nixpkgs version" nixpkgsVersions (nixpkgsName: nixpkgs-pin:
   let pinnedNixpkgsSrc = sources.${nixpkgs-pin};
       # We need this for generic nixpkgs stuff at the right version
       genericPkgs = import pinnedNixpkgsSrc {};
-  in dimension "System" (systems genericPkgs) (systemName: system:
-    let pkgs = import pinnedNixpkgsSrc (nixpkgsArgs // { inherit system; });
-        build = import ./build.nix { inherit pkgs ifdLevel; };
-        platformFilter = platformFilterGeneric pkgs system;
-        compilers = {
-          inherit (pkgs.haskell-nix.compiler) ghc865 ghc883;
-        };
-    in filterAttrsOnlyRecursive (_: v: platformFilter v) {
-      # Native builds
-      # TODO: can we merge this into the general case by picking an appropriate "cross system" to mean native?
-      native = pkgs.recurseIntoAttrs ({
-        inherit (build) tests tools maintainer-scripts maintainer-script-cache;
-        ghc = pkgs.recurseIntoAttrs compilers;
-      } // pkgs.lib.optionalAttrs (ifdLevel >= 1) {
-        iserv-proxy = pkgs.recurseIntoAttrs (
-          pkgs.lib.mapAttrs (ghcName: _:
-            pkgs.ghc-extra-packages."${ghcName}".iserv-proxy.components.exes.iserv-proxy
-          ) compilers);
-      } // pkgs.lib.optionalAttrs (ifdLevel >= 2) {
-        hello = (pkgs.haskell-nix.hackage-package { name = "hello"; version = "1.0.0.2"; }).components.exes.hello;
-      });
-    }
-    //
-    dimension "Cross system" (crossSystems nixpkgsName genericPkgs system) (crossSystemName: crossSystem:
-      # Cross builds
-      let pkgs = import pinnedNixpkgsSrc (nixpkgsArgs // { inherit system crossSystem; });
+  in dimension "GHC version" compilerNixNames (compilerNixName: nixpkgsArgs:
+    dimension "System" (systems genericPkgs) (systemName: system:
+      let pkgs = import pinnedNixpkgsSrc (nixpkgsArgs // { inherit system; });
           build = import ./build.nix { inherit pkgs ifdLevel; };
-      in pkgs.recurseIntoAttrs (pkgs.lib.optionalAttrs (ifdLevel >= 1) {
-        ghc = pkgs.recurseIntoAttrs compilers;
-        # TODO: look into making tools work when cross compiling
-        # inherit (build) tools;
-      } // pkgs.lib.optionalAttrs (ifdLevel >= 2) {
-        remote-iserv = pkgs.recurseIntoAttrs (
-          pkgs.lib.mapAttrs (ghcName: _:
-            pkgs.ghc-extra-packages."${ghcName}".remote-iserv.components.exes.remote-iserv
-          ) compilers);
-        iserv-proxy = pkgs.recurseIntoAttrs (
-          pkgs.lib.mapAttrs (ghcName: _:
-            pkgs.ghc-extra-packages."${ghcName}".iserv-proxy.components.exes.iserv-proxy
-          ) compilers);
-      } // pkgs.lib.optionalAttrs (ifdLevel >= 3) {
-        hello = (pkgs.haskell-nix.hackage-package { name = "hello"; version = "1.0.0.2"; }).components.exes.hello;
+          platformFilter = platformFilterGeneric pkgs system;
+          compilers = {
```

Comment:
I think you can also get rid of this now, since you're now mapping over the compiler versions as a dimension.